### ISSUE TITLE
Android N compability

### ISSUE THREAD

**Issue by tobrun** [mapbox/mapbox-gl-native#5030]
**TL;DR** Android N is coming. This is a WIP ticket trying to summarise the new features introduced in  Android N, link the related Google I/O sessions and determine the impact on our SDK. 

### Features 
This list recaps the content shown in the [`What's new in Android`](https://www.youtube.com/watch?v=B08iLAtS3AQ&list=PLWz5rJ2EKKc8jQTUYvIfqA9lMvSGQWtte) talk:
 - New layout editor:
   - constraint layout
 - Multi-Window mode: 
   - split screen
   - picture in picture mode
   - drag and drop
 - Notifications
 - Quick settings
   - add/remove/reorder items
   - public API
 - Display size: scale all content on screen
 - Multi locale
 - Doze: improved more restricted
 - Project Svelte: job scheduler integration
 - Data Saver: per app restrictions
 - Direct boot
 - Scoped directory access to file system
 - Android N for work
 - ART Runtime improvements
 - Java 8 Language features
 - Lambdas
 - Default and static interface methods
 - Repeating annotations (vs array annotations)
 - Java Audio Latency
 - Renderscript changes
 - OpenGL ES 3.2
 - Vulkan
 - adb shell
 - NDK
 - VR
 - Support library
 - Android instant apps
 
### Google I/O sessions on Android:
This is a list copied from [Android Developer channel](https://www.youtube.com/playlist?list=PLWz5rJ2EKKc8jQTUYvIfqA9lMvSGQWtte). This list was created to filter out sessions (strikethrough) and highlight what I think are useful sessions for us (bolt). The checkable part is only used to track of my own progress. 

- [x] **[What's new in Android](https://www.youtube.com/watch?v=B08iLAtS3AQ&list=PLWz5rJ2EKKc8jQTUYvIfqA9lMvSGQWtte&index=1)**
- [x] [Image compression for Android developers](https://www.youtube.com/watch?v=r_LpCi6DQME&list=PLWz5rJ2EKKc8jQTUYvIfqA9lMvSGQWtte&index=2)
- [x] ~~[Android Pay everywhere: New developments](https://www.youtube.com/watch?v=JQ8w_TbQqVA&list=PLWz5rJ2EKKc8jQTUYvIfqA9lMvSGQWtte&index=3)~~
- [ ] ~~[Games: The Google advantage](https://www.youtube.com/watch?v=5GIk9ABs8ww&list=PLWz5rJ2EKKc8jQTUYvIfqA9lMvSGQWtte&index=4)~~
- [ ] ~~[Beyond payments with Android Pay](https://www.youtube.com/watch?v=Z4iXOmtHTIc&list=PLWz5rJ2EKKc8jQTUYvIfqA9lMvSGQWtte&index=5)~~
- [ ] [Lean and Fast: Putting Your App on a Diet](https://www.youtube.com/watch?v=xctGIB81D2w&list=PLWz5rJ2EKKc8jQTUYvIfqA9lMvSGQWtte&index=6)
- [ ] [Introducing the Awareness API, an easy way to make your apps context aware](https://www.youtube.com/watch?v=37ia7S4Lsv4&list=PLWz5rJ2EKKc8jQTUYvIfqA9lMvSGQWtte&index=7)
- [ ] [Behind the scenes: What’s new in Android accessibility](https://www.youtube.com/watch?v=QlhU0YioJks&list=PLWz5rJ2EKKc8jQTUYvIfqA9lMvSGQWtte&index=8)
- [x] **[What's new in the support library](https://www.youtube.com/watch?v=w45y_w4skKs&list=PLWz5rJ2EKKc8jQTUYvIfqA9lMvSGQWtte&index=9)**
- [ ] ~~[Raiders of the lost app: Google Play secrets to launching and getting discovered](https://www.youtube.com/watch?v=j33M5BqoLuI&list=PLWz5rJ2EKKc8jQTUYvIfqA9lMvSGQWtte&index=10)~~
- [ ] [Designing for driving](https://www.youtube.com/watch?v=JiwiTDXTO78&list=PLWz5rJ2EKKc8jQTUYvIfqA9lMvSGQWtte&index=11)
- [x] **[Multi-Window mode](https://www.youtube.com/watch?v=yEEy_48hoXI&list=PLWz5rJ2EKKc8jQTUYvIfqA9lMvSGQWtte&index=12)**
- [x] ~~[Introducing Google Developer Certification: Become an Associate Android Developer](https://www.youtube.com/watch?v=Yu2oGere_Mc&list=PLWz5rJ2EKKc8jQTUYvIfqA9lMvSGQWtte&index=13)~~
- [ ] [What the Fragment?](https://www.youtube.com/watch?v=k3IT-IJ0J98&list=PLWz5rJ2EKKc8jQTUYvIfqA9lMvSGQWtte&index=14)
- [x] **[Make shinier, faster mobile games with Vulkan](https://www.youtube.com/watch?v=zV_H3QzRhuI&list=PLWz5rJ2EKKc8jQTUYvIfqA9lMvSGQWtte&index=15)**
- [x] **[Android battery and memory optimizations](https://www.youtube.com/watch?v=VC2Hlb22mZM&list=PLWz5rJ2EKKc8jQTUYvIfqA9lMvSGQWtte&index=16)**
- [ ] [Google Cast & Android TV: Building connected experiences for the home](https://www.youtube.com/watch?v=HukWdsRayqQ&list=PLWz5rJ2EKKc8jQTUYvIfqA9lMvSGQWtte&index=17)
- [ ] [Android Auto: The Road Ahead](https://www.youtube.com/watch?v=0Cv1EFhYU2M&list=PLWz5rJ2EKKc8jQTUYvIfqA9lMvSGQWtte&index=18)
- [ ] [Introducing Nearby: Physical proximity within and without apps](https://www.youtube.com/watch?v=Acdu2ZdBaZE&list=PLWz5rJ2EKKc8jQTUYvIfqA9lMvSGQWtte&index=19)
- [x] **[Making Android sensors and location work for you](https://www.youtube.com/watch?v=OEvycEMoLUg&list=PLWz5rJ2EKKc8jQTUYvIfqA9lMvSGQWtte&index=20)**
- [x] **[Android themes & styles demystified](https://www.youtube.com/watch?v=TIHXGwRTMWI&list=PLWz5rJ2EKKc8jQTUYvIfqA9lMvSGQWtte&index=21)**
- [ ] **[What's new in Android security (M and N Version)](https://www.youtube.com/watch?v=XZzLjllizYs&list=PLWz5rJ2EKKc8jQTUYvIfqA9lMvSGQWtte&index=22)**
- [x] **[What's new in Android development tools](https://www.youtube.com/watch?v=csaXml4xtN8&list=PLWz5rJ2EKKc8jQTUYvIfqA9lMvSGQWtte&index=23)**
- [ ] ~~[Best practices in media playback](https://www.youtube.com/watch?v=iIKxyDRjecU&list=PLWz5rJ2EKKc8jQTUYvIfqA9lMvSGQWtte&index=24)~~
- [ ] [What’s new in Android Wear 2.0?](https://www.youtube.com/watch?v=L_Z3gSXGsyI&list=PLWz5rJ2EKKc8jQTUYvIfqA9lMvSGQWtte&index=25)
- [ ] ~~[What's new in Google Play for developers](https://www.youtube.com/watch?v=eGGwReIE61w&list=PLWz5rJ2EKKc8jQTUYvIfqA9lMvSGQWtte&index=26)~~
- [ ] **[The experts' guide to Android development tools](https://www.youtube.com/watch?v=hHnTIMjd1Y8&list=PLWz5rJ2EKKc8jQTUYvIfqA9lMvSGQWtte&index=27)**
- [ ] ~~[Building rich fitness experiences with Google Fit platform and Android Wear](https://www.youtube.com/watch?v=tP8yKNXXGmQ&list=PLWz5rJ2EKKc8jQTUYvIfqA9lMvSGQWtte&index=28)~~
- [ ] **[Advanced Espresso](https://www.youtube.com/watch?v=isihPOY2vS4&list=PLWz5rJ2EKKc8jQTUYvIfqA9lMvSGQWtte&index=29)**
- [ ] ~~[Android Wear 2.0: Making Watch Apps more Standalone](https://www.youtube.com/watch?v=Hw37dxW6q5g&list=PLWz5rJ2EKKc8jQTUYvIfqA9lMvSGQWtte&index=30)~~
- [x] **[Material improvements](https://www.youtube.com/watch?v=EjTJIDKT72M&list=PLWz5rJ2EKKc8jQTUYvIfqA9lMvSGQWtte&index=31)**
- [x] ~~[Your Apps at work](https://www.youtube.com/watch?v=Za0OQo8DRM4&list=PLWz5rJ2EKKc8jQTUYvIfqA9lMvSGQWtte&index=32)~~
- [ ] [Bring Your Android App to Chrome OS](https://www.youtube.com/watch?v=ZLYzX0G0YKQ&list=PLWz5rJ2EKKc8jQTUYvIfqA9lMvSGQWtte&index=33)
- [ ] [Building for billions on Android](https://www.youtube.com/watch?v=vaEV8bNi1Dw&list=PLWz5rJ2EKKc8jQTUYvIfqA9lMvSGQWtte&index=34)
- [ ] ~~[Android Wear 2.0: Building Apps with Material Design](https://www.youtube.com/watch?v=LtD7eJp2ILo&list=PLWz5rJ2EKKc8jQTUYvIfqA9lMvSGQWtte&index=35)~~
- [ ] ~~[Android high-performance audio](https://www.youtube.com/watch?v=F2ZDp-eNrh4&list=PLWz5rJ2EKKc8jQTUYvIfqA9lMvSGQWtte&index=36)~~
- [ ] ~~[Android Wear 2.0: Watch faces and Complications](https://www.youtube.com/watch?v=o-NiVh1YEG0&list=PLWz5rJ2EKKc8jQTUYvIfqA9lMvSGQWtte&index=37)~~
- [x] **[Android NDK performance in an ART world](https://www.youtube.com/watch?v=Ok97X9Z4Si8&list=PLWz5rJ2EKKc8jQTUYvIfqA9lMvSGQWtte&index=38)**
- [x] **[Android Layouts: a new world](https://www.youtube.com/watch?v=sO9aX87hq9c&list=PLWz5rJ2EKKc8jQTUYvIfqA9lMvSGQWtte&index=39)**
- [ ] [Android application architecture: Get ready for the next billion!](https://www.youtube.com/watch?v=70WqJxymPr8&list=PLWz5rJ2EKKc8jQTUYvIfqA9lMvSGQWtte&index=40)
- [ ] ~~[Developing successful games with Google Play](https://www.youtube.com/watch?v=F5HRIT6EAZY&list=PLWz5rJ2EKKc8jQTUYvIfqA9lMvSGQWtte&index=41)~~
- [ ] [RecyclerView ins and outs](https://www.youtube.com/watch?v=LqBlYJTfLP4&list=PLWz5rJ2EKKc8jQTUYvIfqA9lMvSGQWtte&index=42) 
- [ ] [Inclusive design and testing: Making your app accessible](https://www.youtube.com/watch?v=SOZwfQO4rVM&list=PLWz5rJ2EKKc8jQTUYvIfqA9lMvSGQWtte&index=43)
- [ ] ~~[Smarter approaches to a stellar rating on Google Play](https://www.youtube.com/watch?v=p4fjQ112e2I&list=PLWz5rJ2EKKc8jQTUYvIfqA9lMvSGQWtte&index=44)~~

### General Google I/O developer sessions:
This is a curated list of the sessions from the general Google developers channel. The full list can be found [here](https://www.youtube.com/playlist?list=PLOU2XLYxmsILe6_eGvDN3GyiodoV3qNSC). Note that there are also separate lists for [Firebase](https://www.youtube.com/playlist?list=PLl-K7zZEsYLlAyGS6_paVoGJ9YKC7J3NN) and [Chrome](https://www.youtube.com/playlist?list=PLNYkxOF6rcIDz1TzmmMRBC-kd8zPRTQIP).

 - [x] **[Google I/O 2016 - Keynote](https://www.youtube.com/watch?v=862r3XS2YB0&index=1&list=PLOU2XLYxmsILe6_eGvDN3GyiodoV3qNSC)**
 - [ ] [Discover the expanded Material Design motion guidelines](https://www.youtube.com/watch?v=6p3i6H2oGa0&index=20&list=PLOU2XLYxmsILe6_eGvDN3GyiodoV3qNSC)
 - [ ] [Location and Proximity Superpowers: Eddystone + Google Beacon Platform](https://www.youtube.com/watch?v=3nYyApSiSLQ&index=30&list=PLOU2XLYxmsILe6_eGvDN3GyiodoV3qNSC)
 - [ ] [Understand your Place in this world](https://www.youtube.com/watch?v=QR3PIg0RDnk&index=37&list=PLOU2XLYxmsILe6_eGvDN3GyiodoV3qNSC)
 - [ ] [Learning to speak Designer](https://www.youtube.com/watch?v=80hrknwD74Q&index=43&list=PLOU2XLYxmsILe6_eGvDN3GyiodoV3qNSC)
 - [ ] [Principles of mobile app design: Delight users and drive conversions](https://www.youtube.com/watch?v=u7iUoxqKaKU&index=72&list=PLOU2XLYxmsILe6_eGvDN3GyiodoV3qNSC)
 - [ ] [What's New with Project Tango](https://www.youtube.com/watch?v=yvgPrZNp4So&index=75&list=PLOU2XLYxmsILe6_eGvDN3GyiodoV3qNSC)
 - [ ] [Machine learning & art](https://www.youtube.com/watch?v=egk683bKJYU&index=86&list=PLOU2XLYxmsILe6_eGvDN3GyiodoV3qNSC)
 - [ ] [Streamlining developer experiences with the Google Maps APIs](https://www.youtube.com/watch?v=WJY2tGVQDGQ&index=89&list=PLOU2XLYxmsILe6_eGvDN3GyiodoV3qNSC)
 - [ ] [The new Cast SDK](https://www.youtube.com/watch?v=Ij9xM4Velno&index=90&list=PLOU2XLYxmsILe6_eGvDN3GyiodoV3qNSC)
 - [ ] [Machine Learning: Google's Vision](https://www.youtube.com/watch?v=Rnm83GqgqPE&list=PLOU2XLYxmsILe6_eGvDN3GyiodoV3qNSC&index=101)
 - [ ] [3rd Annual Google Security Update](https://www.youtube.com/watch?v=gwbmGvur5VE&list=PLOU2XLYxmsILe6_eGvDN3GyiodoV3qNSC&index=105)
 - [ ] [Election 2016: The big data showdown](https://www.youtube.com/watch?v=tE1j_zzIgCA&index=114&list=PLOU2XLYxmsILe6_eGvDN3GyiodoV3qNSC)
 - [x] [Bridging the physical and digital. Imagine the possibilities. ATAP. - Google I/O 2016](https://www.youtube.com/watch?v=8LO59eN9om4&index=115&list=PLOU2XLYxmsILe6_eGvDN3GyiodoV3qNSC)
 - [ ] [Design Fireside Chat](https://www.youtube.com/watch?v=egsXeMiyV9s&index=121&list=PLOU2XLYxmsILe6_eGvDN3GyiodoV3qNSC)

### To test, validate, research
  - [ ] [Multi Window Support/ PIP mode](https://github.com/mapbox/mapbox-gl-native/issues/5030#issuecomment-220568476)
  - [ ] [Screen off doze mode](https://github.com/mapbox/mapbox-gl-native/issues/5030#issuecomment-220763051)
  - [ ] [OpenGL ES 3.2](https://github.com/mapbox/mapbox-gl-native/issues/5030#issuecomment-220764399)
  - [ ] [Vulkan](https://github.com/mapbox/mapbox-gl-native/issues/5030#issuecomment-220764399)
  - [ ]  [NDK: JNI with ART best practices](https://www.youtube.com/watch?v=Ok97X9Z4Si8&list=PLWz5rJ2EKKc8jQTUYvIfqA9lMvSGQWtte&index=38)
  - [ ] [Test constraint layout](https://www.youtube.com/watch?v=sO9aX87hq9c&list=PLWz5rJ2EKKc8jQTUYvIfqA9lMvSGQWtte&index=39))
  - [ ] [Espresso recorder tests](https://sites.google.com/a/android.com/tools/recent/androidstudio22preview3available)

cc @bleege @zugaldia @cammace @danswick

**Comment by tobrun** [comment]
When having 2 activities visible with multi window support, only the focused activity is in resumed state, while the other one is in paused state. We need to look into if registering for location updates in onResume/onPause should be replaced by onStart/onStop. More [info](https://www.youtube.com/watch?v=yEEy_48hoXI).

**Comment by tobrun** [comment]
<img width="774" alt="screen shot 2016-05-21 at 09 01 58" src="https://cloud.githubusercontent.com/assets/2151639/15447048/5f812f88-1f33-11e6-909e-f35d4c49aa7f.png">

<img width="848" alt="screen shot 2016-05-21 at 08 51 41" src="https://cloud.githubusercontent.com/assets/2151639/15447025/362670f4-1f32-11e6-9206-dc5d4140adc9.png">

<img width="826" alt="screen shot 2016-05-21 at 08 52 50" src="https://cloud.githubusercontent.com/assets/2151639/15447026/3a47cebc-1f32-11e6-82da-1fe3fedc93c2.png">

More info in [battery and memory optimisations](https://www.youtube.com/watch?v=VC2Hlb22mZM) talk.

**Comment by tobrun** [comment]
<img width="792" alt="screen shot 2016-05-21 at 09 29 56" src="https://cloud.githubusercontent.com/assets/2151639/15447133/9ec2c230-1f36-11e6-877d-deceba1e3ac5.png">

Data saver allows to configure per app restrictions, above API allows to check for these.

**Comment by tobrun** [comment]
<img width="739" alt="screen shot 2016-05-21 at 09 36 22" src="https://cloud.githubusercontent.com/assets/2151639/15447146/81bac63c-1f37-11e6-8361-debe771eda79.png">

More information on `The evolution of ART session`.

**Comment by tobrun** [comment]
<img width="779" alt="screen shot 2016-05-21 at 09 46 18" src="https://cloud.githubusercontent.com/assets/2151639/15447190/fd3cfbf8-1f38-11e6-8141-e5cf215bf080.png">

**Comment by tobrun** [comment]
<img width="794" alt="screen shot 2016-05-21 at 09 50 25" src="https://cloud.githubusercontent.com/assets/2151639/15447205/7edcc620-1f39-11e6-8a43-388c3e8abcbb.png">

**Comment by tobrun** [comment]
<img width="777" alt="screen shot 2016-05-21 at 09 51 47" src="https://cloud.githubusercontent.com/assets/2151639/15447209/a90ca168-1f39-11e6-8863-e5f0296ce008.png">

**Comment by zugaldia** [comment]
@tobrun Do you need a device to run some tests? I've got a 25% off discount for a Pixel C (valid until  July 15th) that we could use for this.

**Comment by bleege** [comment]
@tobrun Thanks for collating all the currently known new features expected in Android N.  A suggestion for how to approach this is to first identify things that we know will cause the Mapbox Android SDK to not automatically work in Android N, if any exist at all.  Once we can ensure that the Mapbox Android SDK will run on Android N we can then work on prioritizing new features that will enhance the Mapbox Android SDK across the entire supported range ([currently API 15 and above](https://www.mapbox.com/android-sdk/#supported_android_versions)).

The big thing to keep in mind here is that the number of Android N devices in use will be very small initially compared to other older versions of Android and likely won't move significantly until next year's flagship phones (from Samsung, LG, HTC, etc) go on sale next Spring.  A great example of this is that last year's release of Android Marshmallow is still only at 7.5% market share according to the [Android Dashboard](https://developer.android.com/about/dashboards/index.html).

<img width="776" alt="screen shot 2016-05-21 at 3 59 35 pm" src="https://cloud.githubusercontent.com/assets/989928/15451252/a2f56b32-1f6d-11e6-8fd3-55bd6431402c.png">

**Comment by tobrun** [comment]
> @tobrun Do you need a device to run some tests? I've got a 25% off discount for a Pixel C (valid until July 15th) that we could use for this.

I have both a Pixel C and a Nexus 6P running N-beta, we are good to go from my end 🚀 

> A suggestion for how to approach this is to first identify things that we know will cause the Mapbox Android SDK to not automatically work in Android N, if any exist at all. Once we can ensure that the Mapbox Android SDK will run on Android N we can then work on prioritizing new features that will enhance the Mapbox Android SDK.

💯, my thoughts exactly. 

These are my I created this issue to 
 - identify those issues 
 - create followup issues if items need investigation/testing
 - create a backlog of ideas/features/enhancements
 - filter unuseful content for the team (strikethrough in content list above) 
 - keep track of of viewed sessions. 

> dentify things that we know will cause the Mapbox Android SDK to not automatically work in Android N, if any exist at all.

I'm estimating that there will not be that many but I was able to identify the following:

> Multi window supports works in a way that only one app can be resumed at the same time. This means when an app loses focus it will go into the paused state. For our SDK this is a problem because we register location updates in onResume/onPause. As a result our map will remain visible but will stop receiving location updates. To fix this we are required to bring back the onStart/onStop. All this info is coming from the 15 minute session about [Multi-Window](https://www.youtube.com/watch?v=yEEy_48hoXI&list=PLWz5rJ2EKKc8jQTUYvIfqA9lMvSGQWtte&index=12).

**Comment by tobrun** [comment]
I have been looking into using Espresso test recorder todo UI testing but it seems that touch events on our map are not recorded when. I had high hopes this would solve our UI testing issue but we need to look into other solutions.

**Comment by tobrun** [comment]
I ticketed out #5910 as a code change for this project. Other items are no code changes and will be discussed more internally. Closing for now.
